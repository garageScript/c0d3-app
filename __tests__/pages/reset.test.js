import React from 'react'
import { render, fireEvent, waitFor, act } from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing'
import GET_APP from '../../graphql/queries/getApp'
import UPDATE_PASSWORD from '../../graphql/queries/updatePassword'
import { withTestRouter } from '../../testUtil/withNextRouter'
import ResetPassword from '../../pages/confirm/[token]'

const mockFn = jest.fn()

describe('ResetPassword Page', () => {
  const fillOutResetForm = (getByTestId, password) => {
    const passwordField = getByTestId('password')
    const confirmPasswordField = getByTestId('confirmPassword')

    fireEvent.change(passwordField, {
      target: {
        value: password
      }
    })

    fireEvent.change(confirmPasswordField, {
      target: {
        value: password
      }
    })
  }

  test('Should render confirm success on success', async () => {
    const fakeToken = 'fake token'
    const fakeResetPassword = 'fake reset password'

    const mocks = [
      {
        request: { query: GET_APP },
        result: {
          data: {
            session: null,
            lessons: [],
            alerts: []
          }
        }
      },
      {
        request: {
          query: UPDATE_PASSWORD,
          variables: {
            token: fakeToken,
            password: fakeResetPassword
          }
        },
        result: {
          data: {
            changePw: {
              success: true
            }
          }
        }
      }
    ]

    const tree = withTestRouter(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ResetPassword />
      </MockedProvider>,
      {
        push: jest.fn(),
        query: { token: fakeToken }
      }
    )

    const { container, getByTestId, getByText } = render(tree)

    const submitButton = getByTestId('submit')

    fillOutResetForm(getByTestId, fakeResetPassword)
    fireEvent.click(submitButton)

    await waitFor(() => {
      expect(getByText('Password has been set!')).toBeTruthy()
      expect(getByText('Continue to dashboard')).toBeTruthy()
      expect(container).toMatchSnapshot()
    })
  })

  test('Should render alert on error', async () => {
    const fakeToken = 'fake token'
    const fakeResetPassword = 'fake reset password'

    const mocks = [
      {
        request: { query: GET_APP },
        result: {
          data: {
            session: null,
            lessons: [],
            alerts: []
          }
        }
      },
      {
        request: {
          query: UPDATE_PASSWORD,
          variables: {
            token: fakeToken,
            password: fakeResetPassword
          }
        },
        error: new Error('invalid token')
      }
    ]

    const tree = withTestRouter(
      <MockedProvider mocks={mocks} addTypename={false}>
        <ResetPassword />
      </MockedProvider>,
      {
        push: jest.fn(),
        query: { token: fakeToken }
      }
    )

    const { container, getByTestId, getByText } = render(tree)

    const submitButton = getByTestId('submit')

    fillOutResetForm(getByTestId, fakeResetPassword)
    fireEvent.click(submitButton)

    await waitFor(() => {
      expect(getByText('Link has expired.')).toBeTruthy()
      expect(getByText('Request a new password reset')).toBeTruthy()
      expect(container).toMatchSnapshot()
    })
  })
})
