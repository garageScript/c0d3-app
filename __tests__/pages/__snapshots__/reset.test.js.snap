// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPassword Page Should render alert on error 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/curriculum"
            >
              Curriculum
            </a>
            <a
              class="nav-item nav-link"
              href="https://github.com/garageScript/c0d3-app"
              rel="noopener noreferrer"
              target="_blank"
            >
              Repo
            </a>
            <a
              class="nav-item nav-link"
              href="/docs/learn"
            >
              Learn
            </a>
            <a
              class="nav-item nav-link"
              href="https://discord.gg/c0d3"
              rel="noopener noreferrer"
              target="_blank"
            >
              Community
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
  <div
    class="container-md"
  >
    <div
      class="row mt-5"
    >
      <div
        class="card shadow-sm col-sm-8 col-md-7 col-lg-6 col-xl-6 m-auto px-md-5 border-0"
      >
        <div
          class="card-body text-center pt-5 pb-5"
        >
          <img
            alt="Dark red circle with a exclamation sign inside."
            class="mb-4"
            height="100px"
            src="/assets/curriculum/icons/exclamation.svg"
            width="100px"
          />
          <h1
            class="card-title h2 font-weight-bold mb-5"
          >
            Link has expired.
          </h1>
          <p
            class="card-text"
          />
          <a
            class="btn btn-primary"
            href="/forgotpassword"
          >
            Request a new password reset
          </a>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    © Copyright 2021 GarageScript
  </footer>
</div>
`;

exports[`ResetPassword Page Should render confirm success on success 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/curriculum"
            >
              Curriculum
            </a>
            <a
              class="nav-item nav-link"
              href="https://github.com/garageScript/c0d3-app"
              rel="noopener noreferrer"
              target="_blank"
            >
              Repo
            </a>
            <a
              class="nav-item nav-link"
              href="/docs/learn"
            >
              Learn
            </a>
            <a
              class="nav-item nav-link"
              href="https://discord.gg/c0d3"
              rel="noopener noreferrer"
              target="_blank"
            >
              Community
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
  <div
    class="container-md"
  >
    <div
      class="row mt-5"
    >
      <div
        class="card shadow-sm col-sm-8 col-md-7 col-lg-6 col-xl-6 m-auto px-md-5 border-0"
      >
        <div
          class="card-body text-center pt-5 pb-5"
        >
          <img
            alt="Green circle with a checkmark inside."
            class="mb-4"
            height="100px"
            src="/assets/curriculum/icons/checked.svg"
            width="100px"
          />
          <h1
            class="card-title h2 font-weight-bold mb-5"
          >
            Password has been set!
          </h1>
          <p
            class="card-text"
          />
          <a
            class="btn btn-primary btn-lg mb-3"
            href="/curriculum"
            role="button"
          >
            Continue to dashboard
          </a>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    © Copyright 2021 GarageScript
  </footer>
</div>
`;
